File: packages/ump/android/src/main/java/com/margelo/nitro/umpads/HybridUMPAds.kt
package com.margelo.nitro.umpads

import android.preference.PreferenceManager
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.bridge.ReactApplicationContext
import com.google.android.ump.ConsentDebugSettings
import com.google.android.ump.ConsentInformation
import com.google.android.ump.ConsentRequestParameters
import com.google.android.ump.FormError
import com.google.android.ump.UserMessagingPlatform
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

@DoNotStrip
@Keep
class HybridUMPAds : HybridUMPAdsSpec() {
  private val context: ReactApplicationContext by lazy {
    NitroModules.applicationContext as? ReactApplicationContext
      ?: error("UMP Ads requires a ReactApplicationContext")
  }

  private val consentInformation: ConsentInformation by lazy {
    UserMessagingPlatform.getConsentInformation(context.applicationContext)
  }

  override fun requestConsentInfoUpdate(
    options: UMPConsentRequestOptions?,
  ): Promise<UMPConsentInfo> = Promise.async {
    val activity = context.requireCurrentActivity()

    val params = ConsentRequestParameters.Builder()
      .setTagForUnderAgeOfConsent(options?.tagForUnderAgeOfConsent ?: false)
      .apply {
        val debugGeography = options?.debugGeography
        val testDeviceIds = options?.testDeviceIds
        if (debugGeography != null || !testDeviceIds.isNullOrEmpty()) {
          val debugSettings = ConsentDebugSettings.Builder(activity).apply {
            if (debugGeography != null) setDebugGeography(debugGeography.toNative())
            testDeviceIds?.forEach { addTestDeviceHashedId(it) }
          }.build()
          setConsentDebugSettings(debugSettings)
        }
      }
      .build()

    val deferred = CompletableDeferred<Unit>()
    withContext(Dispatchers.Main) {
      consentInformation.requestConsentInfoUpdate(
        activity,
        params,
        { deferred.complete(Unit) },
        { formError ->
          deferred.completeExceptionally(
            umpException("requestConsentInfoUpdate", formError),
          )
        },
      )
    }
    deferred.await()

    consentInfo()
  }

  override fun loadAndShowConsentFormIfRequired(): Promise<UMPConsentInfo> = Promise.async {
    val activity = context.requireCurrentActivity()

    val deferred = CompletableDeferred<Unit>()
    withContext(Dispatchers.Main) {
      UserMessagingPlatform.loadAndShowConsentFormIfRequired(activity) { formError ->
        if (formError != null) {
          deferred.completeExceptionally(
            umpException("loadAndShowConsentFormIfRequired", formError),
          )
        } else {
          deferred.complete(Unit)
        }
      }
    }
    deferred.await()

    consentInfo()
  }

  override fun showPrivacyOptionsForm(): Promise<UMPConsentInfo> = Promise.async {
    val activity = context.requireCurrentActivity()

    val deferred = CompletableDeferred<Unit>()
    withContext(Dispatchers.Main) {
      UserMessagingPlatform.showPrivacyOptionsForm(activity) { formError ->
        if (formError != null) {
          deferred.completeExceptionally(
            umpException("showPrivacyOptionsForm", formError),
          )
        } else {
          deferred.complete(Unit)
        }
      }
    }
    deferred.await()

    consentInfo()
  }

  override fun showForm(): Promise<UMPConsentInfo> = Promise.async {
    val activity = context.requireCurrentActivity()

    val deferred = CompletableDeferred<Unit>()
    withContext(Dispatchers.Main) {
      UserMessagingPlatform.loadConsentForm(
        context,
        { consentForm ->
          consentForm.show(activity) { formError ->
            if (formError != null) {
              deferred.completeExceptionally(umpException("showForm", formError))
            } else {
              deferred.complete(Unit)
            }
          }
        },
        { formError -> deferred.completeExceptionally(umpException("showForm", formError)) },
      )
    }
    deferred.await()

    consentInfo()
  }

  override fun getConsentInfo(): UMPConsentInfo = consentInfo()

  override fun reset() {
    consentInformation.reset()
  }

  override fun getTCString(): String = sharedPreferences().getString("IABTCF_TCString", "") ?: ""

  override fun getGdprApplies(): Boolean =
    sharedPreferences().getInt("IABTCF_gdprApplies", 0) == 1

  override fun getPurposeConsents(): String =
    sharedPreferences().getString("IABTCF_PurposeConsents", "") ?: ""

  override fun getPurposeLegitimateInterests(): String =
    sharedPreferences().getString("IABTCF_PurposeLegitimateInterests", "") ?: ""

  private fun sharedPreferences() = PreferenceManager.getDefaultSharedPreferences(context)

  private fun consentInfo(): UMPConsentInfo = UMPConsentInfo(
    status = consentInformation.consentStatus.toUMPConsentStatus(),
    isConsentFormAvailable = consentInformation.isConsentFormAvailable,
    privacyOptionsRequirementStatus =
      consentInformation.privacyOptionsRequirementStatus.toUMPStatus(),
    canRequestAds = consentInformation.canRequestAds(),
  )
}

private fun Int.toUMPConsentStatus(): UMPConsentStatus = when (this) {
  ConsentInformation.ConsentStatus.REQUIRED -> UMPConsentStatus.REQUIRED
  ConsentInformation.ConsentStatus.NOT_REQUIRED -> UMPConsentStatus.NOTREQUIRED
  ConsentInformation.ConsentStatus.OBTAINED -> UMPConsentStatus.OBTAINED
  else -> UMPConsentStatus.UNKNOWN
}

private fun ConsentInformation.PrivacyOptionsRequirementStatus.toUMPStatus():
  UMPPrivacyOptionsRequirementStatus = when (this) {
  ConsentInformation.PrivacyOptionsRequirementStatus.REQUIRED ->
    UMPPrivacyOptionsRequirementStatus.REQUIRED
  ConsentInformation.PrivacyOptionsRequirementStatus.NOT_REQUIRED ->
    UMPPrivacyOptionsRequirementStatus.NOTREQUIRED
  else -> UMPPrivacyOptionsRequirementStatus.UNKNOWN
}

private fun UMPDebugGeography.toNative(): Int = when (this) {
  UMPDebugGeography.EEA -> ConsentDebugSettings.DebugGeography.DEBUG_GEOGRAPHY_EEA
  UMPDebugGeography.NOTEEA -> ConsentDebugSettings.DebugGeography.DEBUG_GEOGRAPHY_NOT_EEA
  UMPDebugGeography.DISABLED -> ConsentDebugSettings.DebugGeography.DEBUG_GEOGRAPHY_DISABLED
}

private fun umpException(stage: String, error: FormError) =
  IllegalStateException("UMP $stage failed (${error.errorCode}): ${error.message}")

File: packages/ump/ios/HybridUMPAds.swift
import Foundation
import NitroModules
import UIKit
import UserMessagingPlatform

final class HybridUMPAds: HybridUMPAdsSpec, @unchecked Sendable {
  func requestConsentInfoUpdate(options: UMPConsentRequestOptions?) throws -> Promise<UMPConsentInfo> {
    Promise.async {
      let parameters = RequestParameters()
      parameters.isTaggedForUnderAgeOfConsent = options?.tagForUnderAgeOfConsent ?? false

      if let debugGeography = options?.debugGeography {
        let debugSettings = DebugSettings()
        debugSettings.geography = debugGeography.toNative()
        if let testDeviceIds = options?.testDeviceIds {
          debugSettings.testDeviceIdentifiers = testDeviceIds
        }
        parameters.debugSettings = debugSettings
      }

      // Must run on the main thread/queue — the UMP SDK is not thread-safe
      // and this call can synchronously create UI-related state.
      try await Self.performRequestConsentInfoUpdate(parameters)

      return Self.consentInfo()
    }
  }

  func loadAndShowConsentFormIfRequired() throws -> Promise<UMPConsentInfo> {
    Promise.async {
      let viewController = try await currentViewController()
      // ConsentForm.loadAndPresentIfRequired() creates a WKWebView and must
      // run on the main thread — `await`ing off of `currentViewController()`
      // does not itself guarantee that, so hop explicitly.
      try await Self.performLoadAndPresentIfRequired(from: viewController)
      return Self.consentInfo()
    }
  }

  func showPrivacyOptionsForm() throws -> Promise<UMPConsentInfo> {
    Promise.async {
      let viewController = try await currentViewController()
      try await Self.performPresentPrivacyOptionsForm(from: viewController)
      return Self.consentInfo()
    }
  }

  func showForm() throws -> Promise<UMPConsentInfo> {
    Promise.async {
      let viewController = try await currentViewController()
      try await Self.performShowForm(from: viewController)
      return Self.consentInfo()
    }
  }

  @MainActor
  private static func performRequestConsentInfoUpdate(_ parameters: RequestParameters) async throws {
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
      ConsentInformation.shared.requestConsentInfoUpdate(with: parameters) { error in
        if let error {
          continuation.resume(throwing: error)
        } else {
          continuation.resume()
        }
      }
    }
  }

  @MainActor
  private static func performLoadAndPresentIfRequired(from viewController: UIViewController) async throws {
    try await ConsentForm.loadAndPresentIfRequired(from: viewController)
  }

  @MainActor
  private static func performPresentPrivacyOptionsForm(from viewController: UIViewController) async throws {
    try await ConsentForm.presentPrivacyOptionsForm(from: viewController)
  }

  @MainActor
  private static func performShowForm(from viewController: UIViewController) async throws {
    let form = try await ConsentForm.load()
    try await form.present(from: viewController)
  }

  func getConsentInfo() throws -> UMPConsentInfo {
    Self.consentInfo()
  }

  func reset() throws {
    runOnMain { ConsentInformation.shared.reset() }
  }

  func getTCString() throws -> String {
    UserDefaults.standard.string(forKey: "IABTCF_TCString") ?? ""
  }

  func getGdprApplies() throws -> Bool {
    UserDefaults.standard.bool(forKey: "IABTCF_gdprApplies")
  }

  func getPurposeConsents() throws -> String {
    UserDefaults.standard.string(forKey: "IABTCF_PurposeConsents") ?? ""
  }

  func getPurposeLegitimateInterests() throws -> String {
    UserDefaults.standard.string(forKey: "IABTCF_PurposeLegitimateInterests") ?? ""
  }

  private static func consentInfo() -> UMPConsentInfo {
    UMPConsentInfo(
      status: ConsentInformation.shared.consentStatus.toUMP(),
      isConsentFormAvailable: ConsentInformation.shared.formStatus == .available,
      privacyOptionsRequirementStatus: ConsentInformation.shared.privacyOptionsRequirementStatus.toUMP(),
      canRequestAds: ConsentInformation.shared.canRequestAds
    )
  }
}

private extension ConsentStatus {
  func toUMP() -> UMPConsentStatus {
    switch self {
    case .required: return .required
    case .notRequired: return .notrequired
    case .obtained: return .obtained
    default: return .unknown
    }
  }
}

private extension PrivacyOptionsRequirementStatus {
  func toUMP() -> UMPPrivacyOptionsRequirementStatus {
    switch self {
    case .required: return .required
    case .notRequired: return .notrequired
    default: return .unknown
    }
  }
}

private extension UMPDebugGeography {
  func toNative() -> DebugGeography {
    switch self {
    case .eea: return .EEA
    case .noteea: return .notEEA
    case .disabled: return .disabled
    }
  }
}
